### PR TITLE
Implement gravatar email without public address

### PR DESCRIPTION
**Pull request by iref**
This PR implements  #586. 

I am quite new to elixir and hexpm project, so it is quite possible that some things can be implemented in better way. 

Emails table migration sets user's primary email as her Gravatar email. Is that correct? I know that at the moment public email is used, but based on db schema multiple public emails can exist per users, so it's harder to write generic migration using public email. I would welcome any suggestion how to solve that.

If you have any questions, please don't hesitate to ask. :smile:

### CODE REVIEW

File: lib/hexpm/accounts/users.ex [hunk 1]
```diff
@@ -189,6 +194,52 @@ defmodule Hexpm.Accounts.Users do
...
+          if old_gravatar,
+            do: Multi.update(Multi.new(), :old_gravatar, Email.toggle_gravatar(old_gravatar, false)),
+            else: Multi.new()
+
+        multi
+        |> Multi.update(:new_gravatar, Email.toggle_gravatar(new_gravatar, true))
+        |> audit(audit_data, "email.gravatar", {new_gravatar, nil})
+    end
+  end
+
```
Comment (ericmj) [inline lib/hexpm/accounts/users.ex]: This code is very similar for public, primary and gravatar emails. Maybe we can look at reducing the code duplication?

File: priv/repo/migrations/20171201142544_set_gravatar_on_emails.exs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    execute """
+      UPDATE emails
+      SET gravatar = true
+      WHERE "primary"
+    """
+  end
+
+  def down do
+  end
+end
```
Comment (ericmj) [inline priv/repo/migrations/20171201142544_set_gravatar_on_emails.exs]: Let's combine this into a single migration.

File: test/hexpm/web/views/dashboard_view_test.exs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  test "returns 'none' if gravatar email is not set" do
+    user = %{
+      emails: [build(:email, gravatar: false)]
+    }
+
+    email = DashboardView.gravatar_email_value(user)
+
+    assert email == "none"
+  end
+end
```
Comment (ericmj) [inline test/hexpm/web/views/dashboard_view_test.exs]: 👏

File: lib/hexpm/accounts/users.ex (hunk 1 above)
Comment (iref) [reply lib/hexpm/accounts/users.ex]: Sounds like a good idea. I'll try to simplify it.

**Comment by iref** [comment]
I commited refactoring changes that remove most of the duplication in email flags. I switched ordering of operations in `primary_email`, so we can compose `User.disable_password_reset` with flag switching transaction. Does it matter in this case?

I also fixed views that used public email instead of gravatar.

File: lib/hexpm/web/views/dashboard_view.ex [hunk 4]
```diff
@@ -35,6 +35,19 @@ defmodule Hexpm.Web.DashboardView do
     User.email(user, :public) || "none"
   end
 
+  def gravatar_email_options(user) do
+    emails =
+      user.emails
+      |> Enum.filter(&(&1.verified))
+      |> Enum.map(&{&1.email, &1.email})
+
+    [{"Don't use gravatar", "none"}] ++ emails
```
Comment (ericmj) [inline lib/hexpm/web/views/dashboard_view.ex]: Maybe "Don't show an avatar"?

File: priv/repo/migrations/20171201141936_add_gravatar_to_emails.exs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+defmodule Hexpm.Repo.Migrations.AddGravatarToEmails do
+  use Ecto.Migration
+
+  def up do
+    alter table(:emails) do
+      add :gravatar, :boolean, default: false, null: false
+    end
+    execute """
```
Comment (ericmj) [inline priv/repo/migrations/20171201141936_add_gravatar_to_emails.exs]: One whitespace line between.

File: lib/hexpm/web/views/dashboard_view.ex (hunk 4 above)
Comment (iref) [reply lib/hexpm/web/views/dashboard_view.ex]: Maybe. Seems like better description of empty option.

File: priv/repo/migrations/20171201141936_add_gravatar_to_emails.exs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  def up do
+    alter table(:emails) do
+      add :gravatar, :boolean, default: false, null: false
+    end
+
+    execute """
+      UPDATE emails
+      SET gravatar = true
+      WHERE "primary"
```
Comment (ericmj) [inline priv/repo/migrations/20171201141936_add_gravatar_to_emails.exs]: It was the public email we used for gravatar before.

File: test/hexpm/web/controllers/dashboard_controller_test.exs [hunk 7]
```diff
@@ -273,6 +290,55 @@ defmodule Hexpm.Web.DashboardControllerTest do
...
+    assert get_flash(conn, :info) =~ "gravatar email was changed"
+
+    user = Hexpm.Repo.get!(Hexpm.Accounts.User, c.user.id) |> Hexpm.Repo.preload(:emails)
+    assert Enum.find(user.emails, &(&1.email == "gravatar@mail.com")).gravatar
+    refute Enum.find(user.emails, &(&1.email == "eric@mail.com")).gravatar
+  end
+
+  test "unknown email cannot be gravatar email", c do
+    email = Enum.find(c.user.emails, &(&1.email == "eric@mail.com"))
+    Hexpm.Accounts.Email.toggle_flag(email, :gravatar, true) |> Hexpm.Repo.update
```
Comment (ericmj) [inline test/hexpm/web/controllers/dashboard_controller_test.exs]: `Hexpm.Repo.update!`

**Comment by iref** [comment]
No problem! Thank you for merging. It's been fun problem to work on and to get to know hex.pm project.

**Pull request merged**